src/analysis/processing/qgsalgorithmjoinbylocation.cpp

```diff
@@ -0,0 +1,350 @@
+}
+
+QString QgsJoinByLocationAlgorithm::shortDescription() const
+{
+  return QObject::tr( "Join attributes from one vector layer to another by location." );
+}
+
+QgsJoinByLocationAlgorithm *QgsJoinByLocationAlgorithm::createInstance() const
+{
+  return new QgsJoinByLocationAlgorithm();
+}
+
+
+QVariantMap QgsJoinByLocationAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+
+  mBaseSource.reset( parameterAsSource( parameters, QStringLiteral( "INPUT" ), context ) );
+  if ( !mBaseSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "INPUT" ) ) );
+
+  if ( mBaseSource->hasSpatialIndex() == QgsFeatureSource::SpatialIndexNotPresent )
+    feedback->reportError( QObject::tr( "No spatial index exists for input layer, performance will be severely degraded" ) );
+
+  mJoinSource.reset( parameterAsSource( parameters, QStringLiteral( "JOIN" ), context ) );
+  if ( !mJoinSource )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "JOIN" ) ) );
+
+  mJoinMethod = parameterAsEnum( parameters, QStringLiteral( "METHOD" ), context );
+
+  mFieldNames = parameterAsFields( parameters, QStringLiteral( "JOIN_FIELDS" ), context );
+
+  mPredicates = parameterAsEnums( parameters, QStringLiteral( "PREDICATE" ), context );
+
+  QString prefix = parameterAsString( parameters, QStringLiteral( "PREFIX" ), context );
+
+  if ( mFieldNames.empty() )
+  {
+    mOutFields2 = mJoinSource->fields();
+    mFields2Indices.reserve( mOutFields2.count() );
+    for ( int i = 0; i < mOutFields2.count(); ++i )
+    {
+      mFields2Indices << i;
+    }
+  }
+  else
+  {
+    mFields2Indices.reserve( mFieldNames.count() );
+    for ( const QString &field : mFieldNames )
+    {
+      int index = mJoinSource->fields().lookupField( field );
+      if ( index >= 0 )
+      {
+        mFields2Indices << index;
+        mOutFields2.append( mJoinSource->fields().at( index ) );
+      }
+    }
+  }
+
+  if ( !prefix.isEmpty() )
+  {
+    for ( int i = 0; i < mOutFields2.count(); ++i )
+    {
+      mOutFields2[ i ].setName( prefix + mOutFields2[ i ].name() );
+    }
+  }
+
+  for ( int i = 0; i < mFields2Indices.count(); ++i )
+    mEmptyAttrs << QVariant();
+
+  const QgsFields mOutFields = QgsProcessingUtils::combineFields( mBaseSource->fields(), mOutFields2 );
+
+  mJoinedFeatures.reset( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, mJoinedFeaturesId, mOutFields,
+                                          mBaseSource->wkbType(), mBaseSource->sourceCrs(), QgsFeatureSink::RegeneratePrimaryKey ) );
+
+  if ( parameters.value( QStringLiteral( "OUTPUT" ) ).isValid() && !mJoinedFeatures )
+    throw QgsProcessingException( invalidSinkError( parameters, QStringLiteral( "OUTPUT" ) ) );
+
+  mDiscardNonMatching = parameterAsBoolean( parameters, QStringLiteral( "DISCARD_NONMATCHING" ), context );
+
+  if ( !mDiscardNonMatching )
```

Comment:
These two options are slightly different but independant -- you'll need to remove the above check and create the non_matching sink regardless of the value of DISCARD_NONMATCHING